### PR TITLE
[Core] Priority-based scheduling in async engine

### PR DESCRIPTION
**Pull request by schoennenbeck**
This simply ports the new priority based scheduling logic (https://github.com/vllm-project/vllm/pull/5958)  to the `AsyncLLMEngine`

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by DarkLight1337** [comment]
Please avoid formatting code that isn't related to this PR.

**Comment by schoennenbeck** [comment]
@DarkLight1337 Sorry, my bad. Had format-on-save activated and my personal setup clashes with vllm's codestyle. I'll roll this back.

**Review by DarkLight1337 — APPROVED** [review]
Thanks for extending this to async engine!

**Comment by DarkLight1337** [comment]
It should be in `add_request`, just like the case for regular LLMEngine.

**Comment by schoennenbeck** [comment]
`generate` does call `add_request` internally so should probably support priority as well. The question is whether that should be a top level argument or be part of the `SamplingParams`

**Comment by DarkLight1337** [comment]
> `generate` does call `add_request` internally so should probably support priority as well. The question is whether that should be a top level argument or be part of the `SamplingParams`

I think we can have both.

**Comment by ZJUFangzh** [comment]
In my use case，I want to add priority use tritonserver,  because the priority in tritonserver(dynamic scheduling) is not usefull when it's async to call vllmserver. https://github.com/triton-inference-server/vllm_backend/blob/main/src/model.py#L435

**Comment by schoennenbeck** [comment]
I added priority to all top level APIs of `AsyncLLMEngine` that make sense from my point of view. Feel free to point out if this doesn't fit your use case or if I have the wrong idea about this.

**Comment by DarkLight1337** [comment]
The recent update to transformers 4.45 has resulted in a couple of test failures. You can safely ignore them.

File: vllm/engine/async_llm_engine.py
Comment (njhill) [inline vllm/engine/async_llm_engine.py:420]:
```suggestion
        if priority and not self.scheduler_config.policy == "priority":
```
Negative is also allowed

**Review by njhill — CHANGES_REQUESTED** [review]

File: vllm/engine/async_llm_engine.py
Comment (njhill) [reply vllm/engine/async_llm_engine.py:420]: @schoennenbeck I see that you copied this from LLMEngine ... perhaps you could fix it there too?

Comment (schoennenbeck) [reply vllm/engine/async_llm_engine.py:420]: Sure, will do.

Comment (DarkLight1337) [reply vllm/engine/async_llm_engine.py:420]: You can also merge from `main` to fix the test failures now.

**Comment by sfc-gh-zhwang** [comment]
@schoennenbeck thanks for putting this pr!

File: vllm/engine/async_llm_engine.py
Comment (sfc-gh-zhwang) [inline vllm/engine/async_llm_engine.py:442]: i don't think we are passing `policy="priority"` anywhere, shall we add it to [here](https://github.com/vllm-project/vllm/blob/3b00b9c26c91e9f9ada12975b613555698054e39/vllm/engine/arg_utils.py#L1009) so that user can actually leverage this. Otherwise, how are folks supposed to use this feature? @njhill

File: vllm/engine/async_llm_engine.py
Comment (schoennenbeck) [reply vllm/engine/async_llm_engine.py]: Done and done. I also added the error handling to `add_request` in `AsyncLLMEngine` since it was previously missing.

File: vllm/engine/async_llm_engine.py
Comment (schoennenbeck) [reply vllm/engine/async_llm_engine.py:481]: I think currently the only way of actually using the feature is to manually change the policy after creating the engine.

Comment (sfc-gh-zhwang) [reply vllm/engine/async_llm_engine.py:481]: yes, but shall we allow folks to specify through args? Like other param in the scheduler_config? I feel that's better for wider adoption for this feature.

Comment (schoennenbeck) [reply vllm/engine/async_llm_engine.py:481]: Sorry, I misread your comment. Yes I think this needs to go in the `EngineArgs` or some similar place. However, I think that should be a separate PR.

File: vllm/engine/async_llm_engine.py
Comment (schoennenbeck) [reply vllm/engine/async_llm_engine.py]: @DarkLight1337 The tests still seem to fail for reasons not related to the PR itself.

**Comment by DarkLight1337** [comment]
> The tests still seem to fail for reasons not related to the PR itself.

Yeah, we can force merge if necessary. In any case, we should wait for @njhill's approval.

**Review by njhill — APPROVED** [review]
Sorry for the delay, LGTM thanks @schoennenbeck

File: vllm/engine/async_llm_engine.py
Comment (njhill) [reply vllm/engine/async_llm_engine.py:481]: Yeah it's in SchedulerConfig but still needs to be wired in the EngineArgs to be able to enable it externally. Agree that could be done in a separate PR

**Pull request merged**
